layer/logged_data/system_metrics.py

```diff
@@ -0,0 +1,252 @@
+import pathlib
+import subprocess  # nosec: import_subprocess
+import threading
+import time
+from logging import Logger
+from time import sleep
+from typing import Any, Dict, Optional
+
+import nvsmi  # type: ignore
+import polling  # type: ignore
+import psutil  # type: ignore
+
+from layer.context import get_active_context
+from layer.contracts.logged_data import XCoordinateType
+from layer.logged_data.log_data_runner import LogDataRunner
+
+
+class MetricsCollector:
+    def __init__(self, logger: Logger) -> None:
+        self._logger = logger
+
+    def get_cpu_metrics(self) -> Dict[str, Dict[str, float]]:
+        raise NotImplementedError()
+
+    def get_memory_metrics(self) -> Dict[str, Dict[str, float]]:
+        raise NotImplementedError()
+
+    def get_gpu_metrics(self) -> Dict[str, Dict[str, float]]:
+        metrics: Dict[str, Dict[str, float]] = {}
+        gpu_present = nvsmi.is_nvidia_smi_on_path() is not None
+        if gpu_present:
+            try:
+                gpus = nvsmi.get_gpus()
+            except subprocess.CalledProcessError:
+                self._logger.info(
+                    "Nvidia driver not running despite nvidia-smi being on the path. No GPU stats collected."
+                )
+                return metrics
+            for gpu in gpus:
+                metrics.setdefault("gpu_utilization", {})
+                metrics["gpu_utilization"].update(
+                    {
+                        f"gpu{gpu.id}_utilization_percent": gpu.gpu_util,
+                    }
+                )
+                metrics.setdefault("gpu_memory_utilization", {})
+                metrics["gpu_memory_utilization"].update(
+                    {
+                        f"gpu{gpu.id}_mem_utilisation_percent": round(gpu.mem_util, 2),
+                    }
+                )
+                metrics.setdefault("gpu_memory", {})
+                metrics["gpu_memory"].update(
+                    {
+                        f"gpu{gpu.id}_mem_used_mb": gpu.mem_used,
+                        f"gpu{gpu.id}_mem_allocated_mb": gpu.mem_total,
+                    }
+                )
+
+        return metrics
+
+
+class PsUtilMetricsCollector(MetricsCollector):
+    # cpu
```

Comment:
nit: useless comment (same for `# memory` below)

Reply:
done!